### PR TITLE
Effigy TGS whitelist integration

### PR DESCRIPTION
**Pull request by lessthnthree**
:cl: LT3
server: Effigy whitelist TGS integration
/:cl:

### CODE REVIEW

File: local/code/controllers/subsystem/effigy/whitelist.dm [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+	if(isnull(input_efid))
+		message_admins(span_adminnotice("Attempted to add a player to Effigy whitelist, but no effigy_id provided!"))
+		log_effigy_api("Attempted to add a player to Effigy whitelist, but no effigy_id provided!")
+		return FALSE
+
+	var/ckey_to_match = ckey(input_key)
+	var/requested_link = 0
+	if(isnull(ckey_to_match))
+		message_admins(span_notice("Invalid ckey [ckey_to_match] provided."))
```
Comment (distributivgesetz) [inline local/code/controllers/subsystem/effigy/whitelist.dm:45]:
```suggestion
		message_admins(span_adminnotice("Invalid ckey [ckey_to_match] provided."))
```

File: local/code/controllers/subsystem/effigy/whitelist.dm [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		message_admins(span_adminnotice("Add player [ckey_to_match] to DB whitelist failed!"))
+		log_effigy_api("Add player [ckey_to_match] to DB whitelist failed!")
+		return
+	qdel(query_add_player)
+
+	message_admins(span_info("Add player [ckey_to_match] to DB complete! Verifying link..."))
+	log_effigy_api("Add player [ckey_to_match] to DB complete!")
+	requested_link = SSeffigy.ckey_to_effigy_id(ckey_to_match)
+	if(!requested_link)
+		message_admins(span_notice("Could not find an Effigy ID for ckey [ckey_to_match]!"))
```
Comment (distributivgesetz) [inline local/code/controllers/subsystem/effigy/whitelist.dm:64]:
```suggestion
		message_admins(span_adminnotice("Could not find an Effigy ID for ckey [ckey_to_match]!"))
```
if not applying my first review comment

File: local/code/controllers/subsystem/effigy/whitelist.dm [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+		log_effigy_api("Could not find an Effigy ID for ckey [ckey_to_match]!")
+		return FALSE
+	else
+		message_admins(span_notice("Found Effigy ID [requested_link] for ckey [ckey_to_match]!"))
```
Comment (distributivgesetz) [inline local/code/controllers/subsystem/effigy/whitelist.dm:68]:
```suggestion
		message_admins(span_adminnotice("Found Effigy ID [requested_link] for ckey [ckey_to_match]!"))
```
if not applying my first review comment

File: local/code/controllers/subsystem/effigy/whitelist.dm [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		VALUES (:ckey, :effigy_id, Now(), :round_id, Now(), :round_id, "0", "0", "Player")
+	"}, list("ckey" = ckey_to_match, "effigy_id" = input_efid, "round_id" = GLOB.round_id || null))
+	if(!query_add_player.Execute())
+		qdel(query_add_player)
+		message_admins(span_adminnotice("Add player [ckey_to_match] to DB whitelist failed!"))
+		log_effigy_api("Add player [ckey_to_match] to DB whitelist failed!")
+		return
+	qdel(query_add_player)
+
+	message_admins(span_info("Add player [ckey_to_match] to DB complete! Verifying link..."))
```
Comment (distributivgesetz) [inline local/code/controllers/subsystem/effigy/whitelist.dm:60]:
```suggestion
	message_admins(span_adminnotice("Add player [ckey_to_match] to DB complete! Verifying link..."))
```
if not applying my first review comment

**Review by distributivgesetz — CHANGES_REQUESTED** [review]

File: local/code/controllers/subsystem/effigy/whitelist.dm [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+/datum/tgs_chat_command/effigy_whitelist/Run(datum/tgs_chat_user/sender, params)
+	if(!CONFIG_GET(flag/sql_enabled))
+		return "The Database is not enabled!"
+
+	var/list/all_params = splittext(params, " ")
+	if(all_params.len < 2)
+		return new /datum/tgs_message_content("Requires both a ckey and an effigy_id!")
+
+	var/input_ckey = all_params[1]
+	var/input_efid = all_params[2]
```
[the remark this answers is not in the document]
Comment (lessthnthree) [reply local/code/controllers/subsystem/effigy/whitelist.dm:81]: ah yep, going through tgs skips the validation above

File: code/__HELPERS/_effigy/logging.dm [hunk 6]
```diff
@@ -5,8 +5,10 @@
 /proc/log_subtler(text, list/data)
 	logger.Log(LOG_CATEGORY_EMOTE_SUBTLER, text, data)
 
-/proc/log_effigy_api(text, list/data)
+/proc/log_effigy_api(text, list/data, notify_admins = FALSE, notify_span = "adminnotice")
 	logger.Log(LOG_CATEGORY_EFFIGY_API, text, data)
+	if(notify_admins)
+		message_admins("<span class='[notify_span]'>[text]</span>")
```
Comment (distributivgesetz) [inline code/__HELPERS/_effigy/logging.dm:11]: what a gamer implementation

**Review by distributivgesetz — APPROVED** [review]
looks good to me!

**Pull request merged**